rfcs/0006-transaction.md

```diff
@@ -0,0 +1,299 @@
+
+The good part of this approach is that it's very efficient and simple, and it's free of floating garbage. The bad part is that it may abort the transaction when it's not necessary, and tends to encourage users to set a bigger MemTable size to reduce the probability of the expired transaction.
+
+### Badger
+
+Badger takes a different approach from RocksDB, it provides Serializable isolation level by default. The key idea is to track both read and write operations in the transaction state, and check conflicts with the recent committed transactions tracked in a global Oracle during commit.
+
+Badger maintains a global "Oracle" which tracks the recent committed transactions. When a transaction commits, it should push itself into the Oracle's recent committed transactions list. The Oracle also maintains a sequence number generator, which is used to assign sequence numbers to transactions.
+
+The conflict checking in Badger is as follows:
+
+1. Track the sequence number when transaction started.
+2. Track both read and write keys in the transaction state.
+3. When `commit()` is called, check conflicts with the recent committed transactions in the Oracle which committed after the transaction started.
+4. The conflict check verifies that: the transaction's read and write keys do not conflict with other transactions' write keys
+
+This approach provides Serializable isolation level by detecting both Write-Write conflicts and Read-Write conflicts. It does not rely on MemTable size like RocksDB does, and provides more accurate conflict detection, without worrying about the MemTable size.
+
+The downside is that it requires more memory to track the read keys and recent committed transactions. However, Badger mitigates this by:
+
+1. Using u64 key fingerprints instead of full keys to track read/write keys.
+2. Having an GC mechanism to clean up old committed transactions from the Oracle.
+
+The GC is also straightforward - when all running transactions have sequence numbers greater than a committed transaction's sequence number, that committed transaction can be removed from the Oracle.
+
+## Proposal
+
+### API
+
+The transaction API should be resembles with the non-transactioned API. All the api like `get()`, `put()`, `range()` should also be supported in the transaction API, and these APIs should be aligned with the Snapshot API + WriteBatch API. The only difference is that the transaction API should be able to `commit()` or `rollback()` the changes.
+
+(note: at the time of this RFC, the Snapshot API and WriteBatch API are still not implemented, the transaction feature should be implemented after these APIs are implemented.)
+
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
```

Comment:
It would be good to enumerate the specific APIs for starting a transaction, sending reads/writes to the transaction, and committing it. Mostly what I'd like clarity on is what options we can pass to each of these things, e.g.
- When starting a transaction we should be able to specify a read-level, and probably an isolation level
- When committing a transaction we should be able to specify whether we want to wait for the transaction to be durably applied.

Reply:
https://github.com/slatedb/slatedb/pull/260/files/246378388f65227afb19fcd1c643782b1c4ac6a7#diff-06b7c0e2bb7f360af597d826db9f82b6987ebdbca4cf988381b7a79962493d13R199 here have a introduction about transaction API should contain all the APIs in the Snapshot API and WriteBatch API. it almost contains all the operations in a regular `DB` except the api for creating a transaction.